### PR TITLE
Mizar deploy script

### PR DESCRIPTION
**Pull request by Hong-Chang**

### CODE REVIEW

File: install/deploy_steps.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+source install/common.sh
+
+function deploy_steps:binaries_ready {
+    local is_mizar_production=${1:-0}
+    environment_adaptor:prepare_binary $is_mizar_production
+}
+
+function deploy_steps:environment_ready {
+    common:check_cluster_ready; local is_cluster_ready=$?
```
Comment (phudtran) [inline install/deploy_steps.sh:31]: I noticed that this function is checking the ready status of nodes.
In the deployment of a networking provider such as Mizar, Flannel, etc, we do not care about the status of nodes.

For example, when a cluster is brought up with no networking provider the nodes will always be in "NotReady" state. Once the networking provider is installed, the nodes will transition to  the "Ready" state. 

In the scenario described above, the deployment script will get stuck at this stage and Mizar will never not be installed.

File: deploy-mizar.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+source install/deploy_steps.sh
+
+# Source different adaptor file for different environment. In this way we use same function but do different thing for different environment. 
+common:source_environment_adaptor $mizar_environment
+
+# Execute deploy step by step
+deploy_steps:binaries_ready $is_mizar_production
+deploy_steps:environment_ready $is_mizar_production
+deploy_steps:deploy_mizar $is_mizar_production
+deploy_steps:sanity_check $is_mizar_production
```
Comment (phudtran) [inline deploy-mizar.sh]: Newline

File: etc/deploy/deploy.daemon.dev.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      hostNetwork: true
+      tolerations:
+        - operator: Exists
+          effect: NoSchedule
+      containers:
+        - image: python:3.7
+          name: mizar-daemon
+          command: [ "/bin/bash", "-c", "--", "while true; do mizard; sleep 3; done;" ]
+          securityContext:
+            privileged: true
```
Comment (phudtran) [inline etc/deploy/deploy.daemon.dev.yaml]: New line

File: etc/deploy/deploy.daemon.yaml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          securityContext:
+            privileged: true
+          volumeMounts:
+            - name: mizar
+              mountPath: /home
+      containers:
+        - image: localhost:5000/dropletd:latest
+          name: mizar-daemon
+          securityContext:
+            privileged: true
```
Comment (phudtran) [inline etc/deploy/deploy.daemon.yaml:79]: New line

File: etc/deploy/deploy.initialize.yaml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            type: Directory
+      containers:
+        - image: debian:latest
+          name: mizar-initialize
+          command: [ "/bin/bash", "-c", "--", "while true; do chmod 755/tmp/script-to-run.sh; /tmp/script-to-run.sh; rm -rf /tmp/script-to-run.sh; sleep 3; done;" ]
+          securityContext:
+            privileged: true
+          volumeMounts:
+            - name: mizar
+              mountPath: /home
```
Comment (phudtran) [inline etc/deploy/deploy.initialize.yaml]: New line

File: etc/deploy/deploy.operator.dev.yaml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      hostNetwork: true
+      tolerations:
+        - operator: Exists
+          effect: NoSchedule
+      containers:
+        - image: python:3.7
+          name: mizar-operator
+          command: [ "/bin/bash", "-c", "--", "while true; do ls /etc/luigi/luigi.cfg && kopf run --standalone /var/mizar/mizar/operator.py; sleep 3; done;" ]
+          securityContext:
+            privileged: true
```
Comment (phudtran) [inline etc/deploy/deploy.operator.dev.yaml]: New line

File: etc/deploy/deploy.operator.yaml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      serviceAccountName: mizar-operator
+      hostNetwork: true
+      tolerations:
+        - operator: Exists
+          effect: NoSchedule
+      containers:
+        - image: localhost:5000/endpointopr:latest
+          name: mizar-operator
+          securityContext:
+            privileged: true
```
Comment (phudtran) [inline etc/deploy/deploy.operator.yaml]: New line

File: etc/docker/mizar.Dockerfile [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+# TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+# FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
+# THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+FROM debian
+COPY etc/docker/node-init.sh /
+COPY . /var/mizar
+RUN chmod u+x node-init.sh
```
Comment (phudtran) [inline etc/docker/mizar.Dockerfile]: New line

File: etc/docker/node-init.sh [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+nsenter -t 1 -m -u -n -i mkdir -p /etc/cni/net.d && \
+nsenter -t 1 -m -u -n -i pip3 install --upgrade protobuf && \
+nsenter -t 1 -m -u -n -i pip3 install --ignore-installed /var/mizar/ && \
+nsenter -t 1 -m -u -n -i ln -snf /sys/fs/bpf /bpffs && \
+nsenter -t 1 -m -u -n -i ln -snf /var/mizar/build/bin /trn_bin && \
+nsenter -t 1 -m -u -n -i ln -snf /var/mizar/build/xdp /trn_xdp && \
+nsenter -t 1 -m -u -n -i ln -snf /var/mizar/etc/cni/10-mizarcni.conf /etc/cni/net.d/10-mizarcni.conf && \
+nsenter -t 1 -m -u -n -i ln -snf /var/mizar/mizar/cni.py /opt/cni/bin/mizarcni && \
+nsenter -t 1 -m -u -n -i ln -snf /var/mizar/build/tests/mizarcni.config /etc/mizarcni.config && \
+echo "mizar-complete"
```
Comment (phudtran) [inline etc/docker/node-init.sh]: New line

File: install/environment_adaptors/arktos_up_adaptor.sh [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    sleep 5 # Wait when daemon pod is being created
+    echo "Waiting for daemon pod running. It may cost up to 30 minutes because it needs to setup pip3 modules such as grpcio which needs quite some time for the first time."
+    kubectl wait --for=condition=Ready pod -l job=mizar-daemon --timeout=30m
+    
+    kubectl apply -f etc/deploy/deploy.operator.yaml
+
+    echo "Waiting for Mizar to be up and running."
+    local timeout=60
+    common:execute_and_retry "common:check_mizar_ready" 1 "Mizar is now ready!" "ERROR: Mizar setup timed out after $timeout seconds!" $timeout 1
+   }
```
Comment (phudtran) [inline install/environment_adaptors/arktos_up_adaptor.sh]: New line

File: install/environment_adaptors/default_adaptor.sh [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    kubectl exec $pod_name -- nsenter -t 1 -m -u -n -i mkdir -p /etc/cni/net.d
+    kubectl exec $pod_name -- nsenter -t 1 -m -u -n -i pip3 install --upgrade protobuf
+    kubectl exec $pod_name -- nsenter -t 1 -m -u -n -i pip3 install --ignore-installed /var/mizar/
+    kubectl exec $pod_name -- nsenter -t 1 -m -u -n -i ln -snf /sys/fs/bpf /bpffs
+    kubectl exec $pod_name -- nsenter -t 1 -m -u -n -i ln -snf /var/mizar/build/bin /trn_bin
+    kubectl exec $pod_name -- nsenter -t 1 -m -u -n -i ln -snf /var/mizar/build/xdp /trn_xdp
+    kubectl exec $pod_name -- nsenter -t 1 -m -u -n -i ln -snf /var/mizar/etc/cni/10-mizarcni.conf /etc/cni/net.d/00-mizarcni.conf
+    kubectl exec $pod_name -- nsenter -t 1 -m -u -n -i ln -snf /var/mizar/mizar/cni.py /opt/cni/bin/mizarcni
+    kubectl exec $pod_name -- nsenter -t 1 -m -u -n -i ln -snf /var/mizar/build/tests/mizarcni.config /etc/mizarcni.config
+}
```
Comment (phudtran) [inline install/environment_adaptors/default_adaptor.sh]: New line

File: install/environment_adaptors/k8s_kind_adaptor.sh [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    source install/create_service_account.sh $cwd $user
+
+    source install/deploy_daemon.sh $cwd $user $docker_account
+    source install/deploy_operator.sh $cwd $user $docker_account
+    source install/create_testimage.sh $cwd $user $docker_account
+
+    echo "Waiting for Mizar to be up and running."
+    local timeout=120
+    common:execute_and_retry "common:check_mizar_ready" 1 "Mizar is now ready!" "ERROR: Mizar setup timed out after $timeout seconds!" $timeout 1
+}
```
Comment (phudtran) [inline install/environment_adaptors/k8s_kind_adaptor.sh]: New line

File: install/mizar_deploy_parameters.sh [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+# WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+# TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+# FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
+# THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+MIZAR_HOME="${HOME}/mizar"
+IS_MIZAR_PRODUCTION=1
+MIZAR_ENVIRONMENT="k8s_kind"
```
Comment (phudtran) [inline install/mizar_deploy_parameters.sh]: New line

**Review by phudtran — CHANGES_REQUESTED** [review]
Looks good overall, I added one minor comment.
One minor nitpick, please add a new line to the end of the files.

File: install/deploy_steps.sh (hunk 1 above)
Comment (Hong-Chang) [reply install/deploy_steps.sh:31]: Changed to check cluster status

File: deploy-mizar.sh (hunk 2 above)
Comment (Hong-Chang) [reply deploy-mizar.sh]: Added

File: etc/deploy/deploy.daemon.dev.yaml (hunk 3 above)
Comment (Hong-Chang) [reply etc/deploy/deploy.daemon.dev.yaml]: Added

File: etc/deploy/deploy.daemon.yaml (hunk 4 above)
Comment (Hong-Chang) [reply etc/deploy/deploy.daemon.yaml:79]: Added

File: etc/deploy/deploy.initialize.yaml (hunk 5 above)
Comment (Hong-Chang) [reply etc/deploy/deploy.initialize.yaml]: Added

File: etc/deploy/deploy.operator.dev.yaml (hunk 6 above)
Comment (Hong-Chang) [reply etc/deploy/deploy.operator.dev.yaml]: Added

File: etc/deploy/deploy.operator.yaml (hunk 7 above)
Comment (Hong-Chang) [reply etc/deploy/deploy.operator.yaml]: Added

File: etc/docker/mizar.Dockerfile (hunk 8 above)
Comment (Hong-Chang) [reply etc/docker/mizar.Dockerfile]: Added

File: etc/docker/node-init.sh (hunk 9 above)
Comment (Hong-Chang) [reply etc/docker/node-init.sh]: Added

File: install/environment_adaptors/arktos_up_adaptor.sh (hunk 10 above)
Comment (Hong-Chang) [reply install/environment_adaptors/arktos_up_adaptor.sh]: Added

File: install/environment_adaptors/default_adaptor.sh (hunk 11 above)
Comment (Hong-Chang) [reply install/environment_adaptors/default_adaptor.sh]: Added

File: install/environment_adaptors/k8s_kind_adaptor.sh (hunk 12 above)
Comment (Hong-Chang) [reply install/environment_adaptors/k8s_kind_adaptor.sh]: Added

File: install/mizar_deploy_parameters.sh (hunk 13 above)
Comment (Hong-Chang) [reply install/mizar_deploy_parameters.sh]: Added

**Review by phudtran — APPROVED** [review]
LGTM

**Pull request merged**